garglk/garglk.cpp

```diff
@@ -115,3 +128,99 @@ void garglk_window_get_size_pixels(window_t *win, glui32 *width, glui32 *height)
         *height = hgt;
     }
 }
+
+// Map tuples of (usage, filename, offset) to IDs
+static std::map<std::tuple<glui32, std::string, glui32>, glui32> resource_ids;
+
+// Map IDs to chunks loaded from files
+static std::map<glui32, std::map<glui32, std::vector<unsigned char>>> resource_maps;
+
+// Insert a data chunk into the specified resource map, returning a resource ID
+static glui32 gli_insert_resource(glui32 usage, std::vector<unsigned char> data)
+{
+    auto &map = resource_maps[usage];
+
+    glui32 id = 1;
+    if (!map.empty()) {
+        id = map.rbegin()->first + 1;
+    }
+
+    map.insert({id, std::move(data)});
+
+    return id;
+}
+
+const std::map<glui32, std::vector<unsigned char>> &gli_get_resource_map(glui32 usage)
+{
+    return resource_maps[usage];
+}
+
+glui32 garglk_add_resource_from_file(glui32 usage, const char *filename_, glui32 offset, glui32 len)
+{
+    std::string filename(filename_);
+
+#ifdef _WIN32
+    static const std::vector<std::string> reserved_names = {
+        "con", "prn", "aux", "nul", "com1", "com2", "com3", "com4",
+        "com5", "com6", "com7", "com8", "com9", "lpt1", "lpt2",
+        "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
```

Comment:
Are these needed now that there's a restriction that the file has to be in the same folder as the storyfile?

Reply:
Windows is bizarre and treats these names as special regardless of where they are, at least some of the time.  I wrote a small test program and Windows 10 lets you open an absolute path that ends in "\CON", and treats it like stdin (Windows 11, in my testing, only does this for the filename "CON", not a full path ending in "\CON").

"CON" is a problem because reading from it will hang the program if stdin is connected to anything that will wait for input, and I have no real insight into how various versions of Windows treat the _other_ files for reading, so it's safest to just ban them all.

Reply:
Oh yeah, I have vague memories of seeing that before.

https://superuser.com/questions/86999/why-cant-i-name-a-folder-or-file-con-in-windows

Weird that it also applies to files of those names with any extensions!